### ISSUE TITLE
Design of PyTorch models

### ISSUE THREAD

**Issue by peastman** [deepchem/deepchem#2059]
Now that we're considering adding models based on PyTorch, it seems like a good time to consider the design for them.  We'll want to minimize code duplication, and also ensure the API is as consistent as possible (both between PyTorch based models, and between them and other DeepChem models).  At the same time, we should maintain lots of flexibility and make it easy to use existing PyTorch models from outside.

To get the discussion going, let me offer a design proposal.  This is intended just as a starting point for discussion, so please feel free to suggest changes or alternate designs.

We will create a TorchModel class whose API and feature set are as close as possible to KerasModel.  In place of a `tf.keras.Model`, the constructor will take a `torch.nn.Module`.  Aside from that one difference, the API will be nearly identical.  It will use the same Optimizer, Loss, and Callback objects.  I'm not sure whether all features of KerasModel can be supported, but as far as possible we'll try to do it.

Would that work?  Are there any serious problems we would run into with that approach?

**Comment by rbharath** [comment]
I really like this design! It allows for carry-over of design sense between Keras and PyTorch and should allow for clean docs

**Comment by rbharath** [comment]
One thought that comes to mind is that we probably want to support PyTorch Geometric and DGL as first class citizens. @nd-02110114 has been exploring using DGL/PyTorch for some models and might be able to comment on whether this design makes sense.

At first blush from the DGL docs (https://docs.dgl.ai/tutorials/basics/1_first.html), they use `torch.nn.Module` as the base for their models so I think we should be fine

**Comment by nd-02110114** [comment]
I think this design also work with PyTorch Geometric because many classes which PyG provides are inherited from `torch.nn.module`.

So, I will try to implement TorchModel with similar APIs of KerasModel (Firstly, I will implement the APIs of `fit`, `predict`, `save_checkpoint`)

**Comment by peastman** [comment]
I was planning on implementing this myself.  If you think the design looks good, I can start working on it soon.

**Comment by nd-02110114** [comment]
@peastman 
OK!

**Comment by ncfrey** [comment]
This might not an immediate concern, but is there interest in supporting a high-level API for PyTorch in the same way that DeepChem favors Keras for TensorFlow? I guess it's a more difficult decision because there's no "official" high-level API for PyTorch. 

I'm most familiar with [FastAI](https://github.com/fastai/fastai), but I think it's more useful as a way to quickly get up and running on common deep learning applications, rather than as a general interface to PyTorch. [Ignite](https://github.com/pytorch/ignite) and [Lightning](https://github.com/PyTorchLightning/pytorch-lightning) are two others that seem interesting and more Keras-like. Anyway, maybe just something to keep in mind.

**Comment by nd-02110114** [comment]
I had some experience using Catalyst and PyTorch Lightning. 
Catalyst provides keras-like API for training PyTorch models and I also think Catalyst and PyTorch Lightning and Ignite are options for designing the PyTorch models as Nathan mentioned. 

Catalyst : https://catalyst-team.github.io/catalyst
